File: 1020-number-of-enclaves/1020-number-of-enclaves.cs
public class Solution {
    public int NumEnclaves(int[][] grid) {
        int n=grid.Length;
        int m = grid[0].Length;
        if(n==1 && m==1)
        {
            return 0;
        }
        int[,] visited = new int[n,m];
        Queue<Tuple<int,int>> bfs = new();
        int count = 0;
        for(int i=0; i<n; i++)
        {
            for(int j=0; j<m; j++)
            {
                if(grid[i][j] == 1)
                {
                    count++;
                }
            }
        }
        for(int i=0; i<n; i++)
        {
            if(grid[i][0] == 1)
            {
                bfs.Enqueue(Tuple.Create(i,0));
                visited[i,0]=1;
                count--;
            }
            if(m > 1 &&grid[i][m-1] == 1)
            {
                bfs.Enqueue(Tuple.Create(i,m-1));
                visited[i,m-1]=1;
                count--;
            }
        }
        for(int j=1; j<m-1; j++)
        {
            if(grid[0][j] == 1)
            {
                bfs.Enqueue(Tuple.Create(0,j));
                visited[0,j]=1;
                count--;
            }
            if(n>1 && grid[n-1][j] == 1)
            {
                bfs.Enqueue(Tuple.Create(n-1,j));
                visited[n-1,j]=1;
                count--;
            }
        }
        int[] deltaRow = [1, 0, -1, 0];
        int[] deltaCol = [0, 1, 0, -1];
        while(bfs.Count > 0)
        {
            var top = bfs.Dequeue();
            int row = top.Item1;
            int col = top.Item2;
            for(int k=0; k<4; k++)
            {
                int newRow = row + deltaRow[k];
                int newCol = col + deltaCol[k];
                if(newRow < n && newRow >= 0 && newCol < m && newCol >= 0 && visited[newRow, newCol] == 0 && grid[newRow][newCol] == 1)
                {
                    bfs.Enqueue(Tuple.Create(newRow,newCol));
                    visited[newRow,newCol] = 1;
                    count--;
                }
            }
        }
        return count;
    }
}

File: 1020-number-of-enclaves/1020-number-of-enclaves.java
class Solution {
    public int numEnclaves(int[][] grid) {
        int numRows = grid.length;
        int numColumns = grid[0].length;
        int[][] visited = new int[numRows][numColumns];
        int totalOnes = 0;
        int onesReachableFromBoundary = 0;
        for(int i=0; i<numRows; i++)
        {
            for(int j=0; j<numColumns; j++)
            {
                if(grid[i][j] == 1)
                {
                    totalOnes++;
                    if(((i == 0 || i == numRows-1) || (j == numColumns - 1 || j == 0)) && visited[i][j] == 0)
                    {
                        onesReachableFromBoundary += dfs(i, j, visited, grid, numRows, numColumns);
                    }
                }
            }
        }
        return totalOnes - onesReachableFromBoundary;
    }
    private int dfs(int row, int col, int[][] visited, int[][] grid, int numRows, int numColumns)
    {
        visited[row][col] = 1;
        int size = 1;
        int[][] adjacents = {{0,1}, {1,0}, {-1,0}, {0,-1}};
        for(int[] adjacent : adjacents)
        {
            int newRow = adjacent[0] + row;
            int newCol = adjacent[1] + col;
            if(validatePosition(newRow, newCol, numRows, numColumns, grid) && visited[newRow][newCol] == 0)
            {
                size += dfs(newRow, newCol, visited, grid, numRows, numColumns);
            }
        }
        return size;
    }
    private boolean validatePosition(int row, int col, int numRows, int numColumns, int[][] grid)
    {
        return row < numRows && row >= 0 && col < numColumns && col >= 0 && grid[row][col] == 1;
    }
}